sycl/doc/extensions/SubGroup/SYCL_INTEL_sub_group.asciidoc

```diff
@@ -68,30 +68,25 @@ Providing a generic group abstraction encapsulating the shared functionality of
 
 === Attributes
 
-The +[[intel::reqd_sub_group_size(n)]]+ attribute indicates that the kernel must be compiled and executed with a sub-group of size _n_.  The value of _n_ must be a compile-time integral constant expression.  The value of _n_ must be set to a sub-group size that is both supported by the device and compatible with all language features used by the kernel, or device compilation will fail.  The set of valid sub-group sizes can be queried as described below.
+The +[[intel::sub_group_size(S)]]+ attribute indicates that the kernel must be compiled and executed with a specific sub-group size.  The value of _S_ must be a compile-time integral constant expression.  The kernel should only be submitted to a device that supports that sub-group size (as reported by +info::device::sub_group_sizes+).  If the kernel is submitted to a device that does not support the requested sub-group size, or a device on which the requested sub-group size is incompatible with any language features used by the kernel, the implementation must throw a synchronous exception with the `errc::feature_not_supported` error code from the kernel invocation command.
 
-In addition to device functions, the required sub-group size attribute may also be specified in the definition of a named functor object, as in the example below:
+The +[[intel::named_sub_group_size(NAME)]]+ attribute indicates that the kernel must be compiled and executed with a named sub-group size.  _NAME_ must be one of the following special tokens: +auto+, +primary+.  If _NAME_ is +auto+, the implementation is free to select any of the valid sub-group sizes associated with the device to which the kernel is submitted; the manner in which the sub-group size is selected is implementation-defined.  If _NAME_ is +primary+, the implementation will select the device's primary sub-group size (as reported by the +info::device::primary_sub_group_size+ query) for all kernels with this attribute.
 
-[source, c++]
-----
-class Functor
-{
-  void operator()(item<1> item) [[intel::reqd_sub_group_size(16)]]
-  {
-    /* kernel code */
-  }
-}
-----
+If no sub-group size attribute appears on a kernel, the default behavior is as-if +[[intel::named_sub_group_size(auto)]]+ was specified.  This behavior may be overridden by an implementation (e.g. via compiler flags).
```

Comment:
We should say somewhere that `sub_group_size()` and `named_sub_group_size()` may not both the specified on the same kernel.  This might be a good place to say that.

Reply:
Agreed.  Added in fd5b4072770eb9dc3e85a3065a5cdf70f893a535.

Reply:
So a kernel using `named_sub_group_size(auto)` cannot call a device function in another translation unit?  Why do we need that restriction?

Reply:
I think we need it, at least temporarily, to avoid undefined behavior.

```c++
SYCL_EXTERNAL void foo(); [[intel::sub_group_size(4)]]
SYCL_EXTERNAL void bar(); [[intel::sub_group_size(8)]]
SYCL_EXTERNAL void baz(); [[intel::sub_group_size(auto)]]
SYCL_EXTERNAL void qux(); [[intel::sub_group_size(primary)]]

// Arguably the compiler could choose to compile this kernel with size 4
parallel_for(..., [=](id<1> i) [[intel::sub_group_size(auto)]] {
  foo();
});

// But in this case, "auto" couldn't be evaluated to a single sub-group size
parallel_for(..., [=](id<1> i) [[intel::sub_group_size(auto)]] {
  foo();
  bar();
});

// baz is going to be compiled independently of this kernel
// There's no guarantee that the kernel and function choose the same size for "auto"
parallel_for(..., [=](id<1> i) [[intel::sub_group_size(auto)]] {
  baz();
  /* pretend there's some other code here */
};
```

Since we've agreed that the `SYCL_EXTERNAL` solution needs more discussion, I figured it was better to restrict ourselves only to situations that we can guarantee to support in the future.  Exact matching and no cross-TU magic when handling "auto" seems to be the only case that we all agree could be reasonably supported by different compiler implementations.

Reply:
OK, it's certainly a safe, conservative choice.  We can always lift restrictions later after thinking about it more.

Reply:
I was thinking about this some more.  The default for kernels that have no sub-group size attribute is "auto".  Yet, we are saying that a kernel declared as "auto" cannot call a function defined in a different TU via `SYCL_EXTERN`.  Doesn't this mean that kernels with no sub-group size attribute are forbidden from calling `SYCL_EXTERN` functions?  Surely, that is not our intent?

Reply:
I think you're right, this is what it means.  I'm not sure it's what we'd want in an ideal situation, but I couldn't see a way around it.  If the kernel and function are compiled completely separately and linked later, the only way that linking can occur is if they're both compiled with the same sub-group size.

It might be okay to say that an `auto` kernel calling an `auto` function works, but we'd be forcing all implementations to behave a certain way by doing so.  Using a heuristic to determine the sub-group size for kernels and functions wouldn't be allowed, because they might not match.  There still an open question of what a function pointer to an `auto` function would represent -- would it have to contain some sort of dispatch to different sub-group sizes?  Maybe that's okay too...

Reply:
After our offline discussion yesterday, I've taken this restriction back out in bbeec6a.  Matching `auto` kernels with `auto` functions is something that the current implementation supports, and we shouldn't take that away until we're sure that we want to.  I've left the note about `SYCL_EXTERNAL` in the unresolved questions, and we should revisit this in the context of how sub-groups interact with other extensions.